core/src/main/scala/io/qbeast/spark/index/SparkRevisionFactory.scala

```diff
@@ -58,77 +56,265 @@ object SparkRevisionFactory extends RevisionFactory {
+    else None
   }
 
-  override def createNextRevision(
-      qtableID: QTableID,
-      schema: StructType,
+  /**
+   * Compute the transformation changes for the current operation. The process of creating
+   * transformation changes is as follows:
+   *   1. Compute the Transformations from the input columnStats(if available) and the input
+   *      DataFrame statistics. 2. Check if the new transformations supersede the existing ones.
+   *      3. Make sure no EmptyTransformation is used. If a CDFQuantilesTransformers is present,
+   *      it must be properly setup.
+   * @param transformers
+   *   the updated transformers
+   * @param transformations
+   *   the existing transformations
+   * @param options
+   *   the QbeastOptions
+   * @param row
+   *   the Row containing column statistics computed from the input DataFrame
+   */
+  private[index] def computeTransformationChanges(
+      transformers: IISeq[Transformer],
+      transformations: IISeq[Transformation],
+      options: QbeastOptions,
+      row: Row): IISeq[Option[Transformation]] = {
+    val newTransformations = computeNewTransformations(transformers, options, row)
+    // Create transformationChanges
+    val transformationChanges =
+      if (transformations.isEmpty) newTransformations.map(Some(_))
+      else {
+        // If the input revision has transformations, check if the new transformations supersede them.
+        transformations.zip(newTransformations).map {
+          case (oldT, newT) if oldT.isSupersededBy(newT) => Some(oldT.merge(newT))
+          case _ => None
+        }
+      }
+    // Make sure no EmptyTransformation is used.
+    // If a CDFQuantilesTransformers is present, it must be properly setup
+    transformers.zip(transformationChanges).foreach {
+      case (t, Some(_: EmptyTransformation)) =>
+        var msg = s"Empty transformation for column ${t.columnName}."
+        if (t.isInstanceOf[CDFQuantilesTransformer]) {
+          msg += s" The following must be provided to use QuantileTransformers: ${t.stats.statsNames.head}."
+        }
+        throw AnalysisExceptionFactory.create(msg)
+      case _ =>
+    }
+    transformationChanges
+  }
+
+  /**
+   * Compute the new transformations from the input DataFrame statistics and the columnStats(if
+   * available) and merge them.
+   * @param transformers
+   *   the transformers
+   * @param options
+   *   the QbeastOptions containing the columnStats
+   * @param row
+   *   the Row containing column statistics computed from the input DataFrame
+   */
+  private[index] def computeNewTransformations(
+      transformers: IISeq[Transformer],
       options: QbeastOptions,
-      oldRevisionID: RevisionID): Revision = {
-    val revision = createNewRevision(qtableID, schema, options)
-    revision.copy(revisionID = oldRevisionID + 1)
+      row: Row): IISeq[Transformation] = {
+    val transformationsFromColumnsStats = {
```

Comment:
I would divide this code in three:

1. computeNewTransformationsFromColumnStats
2. computeNewTransformationsFromDataFrame
3. mergeTransformations

In this way, we could encapsulate the merging and add reusable functionalities for 1 and 2

what do you think?

Reply:
Done